src/goto-symex/build_goto_trace.cpp

```diff
@@ -286,6 +294,19 @@ void build_goto_trace(
   for(auto &t_it : time_map)
     goto_trace.steps.splice(goto_trace.steps.end(), t_it.second);
 
+  // cut off the trace at the desired end or after the first failed
```

Comment:
It's going to cut them off at the specified end, which will be set individually for each goal.